src/frontend/src/utils/condition.rs

```diff
@@ -239,6 +240,77 @@ impl Condition {
         .unwrap()
     }
 
+    /// generate range scans from each arm of `OR` clause and merge them
+    /// Currently, only support equal type range scans.
+    /// Keep in mind that range scans can not be overlap, otherwise duplicate rows will occur.
+    fn disjunctions_to_scan_ranges(
+        order_column_ids: &[usize],
+        num_cols: usize,
+        disjunctions: Vec<ExprImpl>,
+    ) -> Result<Option<(Vec<ScanRange>, Self)>> {
+        let disjunctions_result: Result<Vec<(Vec<ScanRange>, Self)>> = disjunctions
+            .into_iter()
+            .map(|x| {
+                Condition {
+                    conjunctions: to_conjunctions(x),
+                }
+                .split_to_scan_ranges(order_column_ids, num_cols)
+            })
+            .collect();
+
+        // if any arm of `OR` clause fail, bail out.
+        let disjunctions_result = disjunctions_result?;
+
+        // if all arm of `OR` clause scan ranges are simply equal condition type, merge all
+        // of them.
+        let all_equal = disjunctions_result
+            .iter()
+            .all(|(scan_ranges, other_condition)| {
+                other_condition.always_true()
+                    && scan_ranges
+                        .iter()
+                        .all(|x| !x.eq_conds.is_empty() && is_full_range(&x.range))
+            });
+
+        if all_equal {
+            // think about the case (a = 1) or (a = 1 and b = 2)
+            // we should only keep the large one range scan a = 1, because a = 1 is overlap with (a
+            // = 1 and b = 2)
```

Comment:
```suggestion
            // Think about the case (a = 1) or (a = 1 and b = 2).
            // We should only keep the large one range scan a = 1, because a = 1 overlaps with
            // (a = 1 and b = 2).
```